### DISCUSSION TITLE
MongoDB의 도입 필요성 및 시기에 관해서 논의할 것이 있습니다!

### DISCUSSION THREAD

**Discussion by thornscript** [GU-99/grow-up-pms#4] [개발 토론]
## 🚀 NoSQL의 도입 필요성
기능 명세서를 보면서 ERD를 짜며 느낀건데 개인적으로 느끼기에는 마크다운 포맷을 저장하기 위해서 NoSQL의 도입을 하는 것에 의문이 들었습니다! 제가 기획 단계에서 참여하지 않기도 하고, 실무 경험이 그렇게 많지 않아 미처 고려하지 못한 부분이 있을 수도 있습니다 🥲 

- 일정(project_schedules) 테이블의 스키마가 빈번하게 변경될 여지가 적음
- 마크다운 같은 간단한 문서의 경우에는 MariaDB/PostgreSQL와 같은 전통적인 관계형 데이터베이스에서도 처리할 수 있음

예를 들어서, 권한(team_permissions, project_permissions)의 경우에는 디스코드나 깃허브처럼 권한 구조가 복잡하고 서비스가 발전함에 따라서 추가적인 권한이 생길 수도 있는 경우에는 NoSQL로 관리해보는 것도 괜찮은 것 같다고 생각이 들었습니다.

```json
{
  "_id": "channel1",
  "serverId": "server1",
  "name": "General",
  "permissions": [
    {
      "roleId": "role1",
      "permissions": ["READ_MESSAGES", "SEND_MESSAGES", "DELETE_MESSAGES", "MANAGE", "ADD_REACTIONS", "MANAGE_CHANNELS", "MANAGE_ROLES", "KICK_MEMBERS", "BAN_MEMBERS", "INVITE_MEMBERS"]
    },
    {
      "roleId": "role2",
      "permissions": ["READ_MESSAGES", "SEND_MESSAGES", "ADD_REACTIONS"]
    },
    {
      "roleId": "role3",
      "permissions": ["READ_MESSAGES"]
    }
  ]
}
```

## 💡 개인적으로 생각한 전략
처음에는 MariaDB를 사용했다가, 필요에 따라서 MongoDB로 마이그레이션 하는 전략을 생각했습니다. 이렇게 하면 초기 개발 속도를 높이고 리스크를 줄일 수 있으며, '서비스가 향후 성장하거나 데이터 처리 요구사항이 달라짐에 따라서 적절한 시점에 NoSQL로 마이그레이션하면 어떨까?'라는 생각이 들었거든요! 🐱

**Comment by KimJayhyun** [comment]
## 🤔 재현의 생각 
### 1. 기획 단계에서 Markdown을 NoSQL(MongoDB)에 저장할려고 기획한 이유
- 공부목적
- MD 형식의 Text 길이가 Max Length를 넘을까봐
![image](https://github.com/GU-99/grow-up-pms/assets/79262624/f2f0b509-137b-40cf-907b-b68f688c4b1f)

### 2. 권한에 따른 MongoDB 사용 이유에 대한 생각

- **공부 목적**으로 권한을 MongoDB에 적재하는 것은 의미가 있음
- 허나 여러 권한이 생긴다는 점에서 아래 의문이 생김
    - 추가적인 권한이 생긴다는 것은 사용자가 임의로 권한을 생성하는 것이 아닌 개발자가 개발한 권한 내에서 설정을 해야 함.
    - 예시로 언급하신 것과 같이 `"READ_MESSAGES", "SEND_MESSAGES", "DELETE_MESSAGES` 처럼 권한을 구별하고 RDB로 관리하는 것이 더 간편하지 않을까 생각함.
    
    ```
    ROLE_ID  | ROLE_NAME
    
    1-0 | MESSAGE_ALL
    1-1 | MESSAGE_READ
    1-2 | MESSAGE_SEND
    ```

### 3. 우리가 가져갈 전략

- 말씀하신 전략대로 먼저 mariadb를 활용하여 초기 개발 속도를 높혀 서비스를 완성시킨 후,
필요하거나 유용하다고 생각되는 Entity를 MongoDB로 마이그래이션을 하는 것이 좋을거 같습니다!

**Comment by thornscript** [comment]
@KimJayhyun 사용자 입장에서 뭔가 커스텀한 권한이 추가된다는 의미는 아니었고 서비스가 확장됨에 따라서 기능이 추가되고 그 기능에 대한 권한이 추가되는 경우를 말하는 것이었습니다! 😺  권한 체계가 디스코드나 깃허브처럼 복잡한 경우에는 한 번 생각해봄직 하다 싶어서 한 번 말씀드려봤습니다...!

전략은 전의 논의에서 정해졌고 그대로 가면 좋을 것 같습니다!
